jira/resilientsession.py

```diff
@@ -113,11 +113,10 @@ def __recoverable(
                 f"Got ConnectionError [{response}] errno:{response.errno} on {request} {url}\n{vars(response)}\n{response.__dict__}"
             )
         if isinstance(response, Response):
-            if response.status_code in [502, 503, 504, 401]:
-                # 401 UNAUTHORIZED still randomly returned by Atlassian Cloud as of 2017-01-16
+            if response.status_code in [429, 502, 503, 504]:
+                # 401 UNAUTHORIZED still randomly returned by Atlassian Cloud as of 2017-01-16 and are handled by jira.JiraCookieAuth.handle_401
                 msg = f"{response.status_code} {response.reason}"
-                # 2019-07-25: Disabled recovery for codes above^
-                return False
+                return True
```

Comment:
this seems to do a lot more than handle the 429 error code, as we are now returning True for a bunch of other error codes.

Additionally, this is in a short circuit clause that returns True... It doesn't actually reach the `time.sleep(delay)` code !

Can I make a few suggestions?
1. We use a separate elif clause for 429, that does not have a short circuit return value. i.e. it just does some logging and has a comment explaining what it is trying to do. For example: "handle the rate limit code".
2. Within that elif clause we have a `logging.warning` message that we have been rate limited, and we include the information in the headers as found in the #925 reference link, i.e. `X-RateLimit-Remaining` 
3. The warning message also includes a suggestion to add the user to the exemption list if they are hitting this too often.
4. We add a debug message to this exist clause that actually logs the msg, currently the `msg` variable is set but then not used...


### For any future readers

Really apologise for the state of this file... (using the root logger for logging too...). The implementation really isn't great

I'm going to try and raise a PR that moves the code towards the suggestion here:
https://stackoverflow.com/questions/23267409/how-to-implement-retry-mechanism-into-python-requests-library/23271433#comment58969710_23271433

Reply:
Thanks for the constructive feedback, you raised some good points! Let me try to address them one by one

1. Rather than adding a new elif clause, can I suggest to just get rid of the other 5xx status codes and keep retries only for the 429 code? I find if confusing for readers without context to see them handled in a separate clause, that is functionally equivalent to handling them in the non-retriable codes execution path.
2-3-4. Sure will do that.

Reply:
Yep good shout, to be honest I got confused reading that why your test passed for the 502 error... missed the `not` in the `elif`